src/wallet/api/wallet2_api.h

```diff
@@ -1144,6 +1149,545 @@ struct Wallet
+    * __________ Approach 1 __________
+    * param: multisig_ptxs -
+    * __________ Approach 2 __________
+    * param: multisig_sign_data -
+    * ________________________________
+    * return : true if enote is frozen, else false
+    * note: sets status error on fail
+    */
+    virtual bool isFrozen(std::size_t idx) const = 0;
+    virtual bool isFrozen(const std::string &key_image) const = 0;
+    virtual bool isFrozen(const PendingTransaction &multisig_ptxs) const = 0;
+    virtual bool isFrozen(const std::string multisig_sign_data) const = 0;
+    // QUESTION : Should we add the following function, which returns strings containing the key image for every frozen enote? I think it can be useful
+//    virtual std::vector<std::string> getAllFrozenEnotes() const = 0;
+    // TODO : Would this better fit in one of `Subaddress`/`SubaddressAccount` classes?
+    /**
+    * brief: createOneOffSubaddress - create a subaddress for given index
+    * param: account_index - major index
+    * param: address_index - minor index
+    */
+    virtual void createOneOffSubaddress(std::uint32_t account_index, std::uint32_t address_index) = 0;
+    /**
+    * brief: isDeprecated - check if wallet file format is deprecated
+    * return: true if wallet file is old format (before becoming JSON), else false
+    */
+    virtual bool isDeprecated() const = 0;
+    /**
+    * brief: hasUnknownKeyImages - check if any enotes have missing key images
+    * return: true if any stored enote is missing its key image, else false
+    */
+    virtual bool hasUnknownKeyImages() const = 0;
+    /**
+    * brief: rewrite - rewrite the wallet file for wallet update
+    * param: wallet_name - name of the wallet file (should exist)
+    * param: password - wallet password
+    * note: sets status error on fail
+    */
+    virtual void rewrite(const std::string &wallet_name, const std::string &password) = 0;
+    // QUESTION : Should we change this function from the current behavior in wallet2, so `wallet_name` is just the name of the new wallet instead of changing the `m_wallet_file` for the current wallet?
+    /**
+    * brief: writeWatchOnlyWallet -  create a new watch-only wallet file with view keys from current wallet
+    * param: wallet_name - name of the current wallet file
+    * param: password - password for new watch-only wallet
+    * outparam: new_keys_file_name - wallet_name + "-watchonly.keys"
+    * note: sets status error on fail
+    */
+    virtual void writeWatchOnlyWallet(const std::string &wallet_name, const std::string &password, std::string &new_keys_file_name) = 0;
+    /**
+    * brief: balancePerSubaddress - get balance for each subaddress in given account
+    * param: index_major - account index
+    * param: strict -
+    * return: [minor subaddress index: amount, ... ]
+    */
+    virtual std::map<std::uint32_t, std::uint64_t> balancePerSubaddress(std::uint32_t index_major, bool strict) const = 0;
```

Comment:
Is this "essential", or is a client only a loop away of summing such things itself by using some call that returns subaddresses and info about them?

Reply:
On first sight it seems the API already can give you the total balance and unlocked balance:
1. from [SubaddressAccountRow](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/wallet/api/wallet2_api.h#L310-L311)
2. from [Wallet](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/wallet/api/wallet2_api.h#L616-L629), this can give you both the total and per account balance.
But I don't see anything for per subaddress balance.

Reply:
Actually since `wallet2::balance_per_subaddress()` (found [here](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/wallet/wallet2.cpp#L6996)) loops over `m_transfers`, `m_unconfirmed_txs` (if not `strict`) and `m_unconfirmed_payments` and `wallet2::unlocked_balance_per_subaddress()` just loops over `m_transfers`, I think a client should be able to do this on their own, once we figured out how to replace those wallet2 members with `EnoteDetails` and `TransferInfo` in the API.
Similar to how simplewallet uses `wallet2::get_transfers()` to get per subaddress balance [here](https://github.com/monero-project/monero/blob/a1dc85c5373a30f14aaf7dcfdd95f5a7375d3623/src/simplewallet/simplewallet.cpp#L8885-L8906).

But I'm thinking subaddress balance is the "lowest level" of summed balances and it would be easier if we provide that feature for clients, so they could implement account balance as the sum of the accounts subaddress balances and total wallet balance as the sum of account balances. On the other hand, as mentioned, we already have account balance and total balance in the API and removing those would not make sense either, so probably I'm just rambling. More opinions are welcome. Should we add per subaddress balance or not?

Reply:
Removed balance per subaddress.